### PR TITLE
feat: v1 streaming protocol

### PR DESCRIPTION
**Pull request by davidbp**

### CODE REVIEW

File: docarray/array/mixins/io/binary.py [hunk 1]
```diff
@@ -22,17 +22,18 @@ def load_binary(
...
+        return_iterator: bool = False,
     ) -> 'T':
         """Load array elements from a LZ4-compressed binary file.
 
         :param file: File or filename or serialized bytes where the data is stored.
         :param protocol: protocol to use
         :param compress: compress algorithm to use
         :param _show_progress: show progress bar, only works when protocol is `pickle` or `protobuf`
-
+        :param return_iterator: returns an iterator over the DocumentArray.
```
Comment (JoanFM) [inline docarray/array/mixins/io/binary.py:33]: what happens if protocol is another? why does protocol affect here?

File: docarray/array/mixins/io/binary.py [hunk 2]
```diff
@@ -41,37 +42,100 @@ def load_binary(
...
-                dap.ParseFromString(d)
+            return cls.from_protobuf(dap)
+        elif protocol == 'pickle-array':
+            return pickle.loads(d)
 
-                return cls.from_protobuf(dap)
-            elif protocol == 'pickle-array':
-                return pickle.loads(d)
+        # Binary format for streaming case
+        else:
```
Comment (JoanFM) [inline docarray/array/mixins/io/binary.py:107]: I am not fan of this `else`. We should perhaps name it

File: docarray/array/mixins/io/binary.py [hunk 3]
```diff
@@ -141,22 +208,40 @@ def to_bytes(
...
 
                         track = lambda x: _track(x, description='Serializing')
                     else:
                         track = lambda x: x
 
+                    # V1 DocArray streaming serialization format
+                    # | 1 byte | 8 bytes | 4 bytes | variable | 4 bytes | variable ...
+
+                    # 1 byte (uint8)
+                    version_byte = b'\x01'
```
Comment (JoanFM) [inline docarray/array/mixins/io/binary.py:230]: this version should come from somewhere else

Comment (JoanFM) [reply docarray/array/mixins/io/binary.py:230]: And should be used to validate when loading that `version` is not compatible

File: tests/unit/array/test_from_to_bytes.py [hunk 4]
```diff
@@ -38,7 +40,7 @@ def test_to_from_bytes(target_da, protocol, compress, ndarray_val, is_sparse):
     assert len(da2) == len(target_da)
 
     target_da.embeddings = ndarray_val
-    target_da.tensors = ndarray_val
+    target_da.blobs = ndarray_val
```
Comment (JoanFM) [inline tests/unit/array/test_from_to_bytes.py:43]: these are tensors, bad merge

File: docarray/array/mixins/io/binary.py (hunk 1 above)
Comment (davidbp) [reply docarray/array/mixins/io/binary.py:33]: Depending on the procol loading changes (as it does in the `main` branch, the behaviour is kept the same)

File: tests/unit/array/test_from_to_bytes.py (hunk 4 above)
Comment (davidbp) [reply tests/unit/array/test_from_to_bytes.py]: updated

File: docarray/array/mixins/io/binary.py (hunk 2 above)
Comment (davidbp) [reply docarray/array/mixins/io/binary.py:107]: ok, making explicit the elif

File: docarray/array/mixins/io/binary.py [hunk 5]
```diff
@@ -22,17 +22,18 @@ def load_binary(
         protocol: str = 'pickle-array',
         compress: Optional[str] = None,
         _show_progress: bool = False,
+        return_iterator: bool = False,
```
Comment (hanxiao) [inline docarray/array/mixins/io/binary.py:25]:
```suggestion
        streaming: bool = False,
```

Comment (hanxiao) [reply docarray/array/mixins/io/binary.py:25]: `T` -> `Union['DocumentArray', Generator['Document', ...]]`

File: docarray/array/mixins/io/binary.py (hunk 1 above)
Comment (hanxiao) [inline docarray/array/mixins/io/binary.py:33]:
```suggestion
        :param streaming: returns an generator over Document.
```

File: docarray/array/mixins/io/binary.py [hunk 6]
```diff
@@ -41,37 +42,100 @@ def load_binary(
             file_ctx = open(file, 'rb')
         else:
             raise ValueError(f'unsupported input {file!r}')
+        if return_iterator:
+            return cls._load_binary_stream(
```
Comment (hanxiao) [inline docarray/array/mixins/io/binary.py:46]: `yield from ...`

File: docarray/array/mixins/io/binary.py [hunk 7]
```diff
@@ -41,37 +42,100 @@ def load_binary(
...
+                file_ctx, protocol=protocol, compress=compress
+            )
+        else:
+            return cls._load_binary_all(file_ctx, protocol, compress, _show_progress)
```
Comment (hanxiao) [inline docarray/array/mixins/io/binary.py:50]:
```python
DocumentArray(yield from ...)

```

File: docs/fundamentals/documentarray/serialization.md [hunk 8]
```diff
@@ -226,10 +211,6 @@ docs {
 
 ## From/to list
 
-```{important}
-This feature requires `protobuf` or `pydantic` dependency. You can do `pip install "docarray[full]"` to install it.
-```
-
 Serializing to/from Python list is less frequently used for the same reason as `Document.to_dict()`: it is often an intermediate step of serializing to JSON. You can do:
```
Comment (hanxiao) [inline docs/fundamentals/documentarray/serialization.md:215]: please revert back

**Review by hanxiao — APPROVED** [review]

**Pull request merged**